packages/core/http/core-http-server/src/router/scratch_pad.ts

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { schema } from '@kbn/config-schema';
+import { RequestHandler } from './request_handler';
+import { RequestHandlerContextBase } from './request_handler_context';
+import { RouteConfig, RouteMethod } from './route';
+import { IRouter } from './router';
+import { RouteValidatorFullConfig } from './route_validator';
+
+/**
+ * Constituents of a route in Kibana
+ *
+ * 1. Route method: put, post, get
+ * 2. Path
+ * 3. Validation
+ * 4. Extra configuration like auth and timeout
+ * 5. Handler function
+ */
+
+// Current API
+
+const router: IRouter = {} as any;
+
+// 1
+router.get(
+  {
+    path: '/api/my-plugin/my-route', // 2
+    validate: {}, // 3
+    options: {}, // 4
+  },
+  // 5
+  async (ctx, req, res) => {
+    return res.ok();
+  }
+);
+
+/**
+ * ===================== First proposal =====================
+ * We ask consumers to restate all consituents for each route.
+ */
+
+type Version = '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | '10';
+
+type NewRouteRegistrar1<
+  P,
+  Q,
+  B,
+  Context extends RequestHandlerContextBase = RequestHandlerContextBase,
+  Method extends RouteMethod = 'get'
+> = (
+  route: RouteConfig<P, Q, B, Method> & { version: Version },
+  handler: RequestHandler<P, Q, B, Context, Method>
```

Comment:
Have you given any thought to how we're going to make the version header actionable? Luke and I were briefly chatting about extending context

Reply:
Yeah that could make sense!

No strong opinions here, I was think of a way to tie validation+handler to a version, so that's the function of passing the `version` value in here. I'm not convinced this is the correct API. I intend to document a few different approaches -- very much WIP.